LuatixHQ/opencti
# [backend/frontend] Add user and group confidence level (#4304)

### Proposed changes
This PR addresses the following issues:
* Closes #4304
* Closes #4901
* Closes #5692  

Have a look at each issue description for details.

### Code review highlights
The key points to review are backend additions and changes, the model definition.
A lot of stuff in the frontend besides the core job (refactor in TS, solving react warnings) make the review difficult, I'm sorry.

### How can you test this PR?
Here are some scenarios to test:

* start the platform on this branch -> you shall see the critical alert dialog, inviting you to edit all your groups

* Create a user without groups -> they should be created with default groups
* Create a user with groups -> they should be created with the specified groups, and not the default group

* Create a user without a confidence level and no group 
  * it should be displayed as in error in the various view (list, details, list of users in a group details view) -> icon + tooltip
  * edit the user : the message above the confidence level field shall be explicit (no group to inherit from)
  * edit the user, tab groups : the confidence of each group shall be displayed, with red icon and tooltip if not defined
  * edit the user : set a group in the groups tab, go back to overview -> the message above the confidence level field shall be explicit (inherited value)
  * edit the user : set the user's confidence level -> -> the message above the confidence level field shall be explicit (user value)
* Create a user without a confidence level and no group -> it should be displayed as in error in the various view (list, details, list of users in a group details view)
* Create a user with a confidence level -> it should be displayed in the various view (list, details, list of users in a group details view) -> tooltip shall state it's coming from user
* Create a user without a confidence level but with a group -> it should be displayed in the various view (list, details, list of users in a group  details view) -> tooltip shall state it's coming from a specific group - link clickable in tooltip

* Create a group, you must select a confidence level -> value is then displayed correctly in list and details
* Open a group without confidence level -> explicit alert in details -> disappears when edited

* all groups are set with a confidence level -> refresh page -> no more alert dialog


If you wish to artificially nuke your group confidence level for testing purpose, a GQL query on playground is possible (there is no validation implemented yet - see #5696).

```graphql
mutation GroupEditionOverviewFieldPatchMutation(
  $id: ID!
  $input: [EditInput]!
) {
  groupEdit(id: $id) {
    fieldPatch(input: $input) {
      id
      group_confidence_level {
        max_confidence
      }
    }
  }
}
```
```json
{
  "id": "4b762ee9-b2da-49fe-99d7-dfd6b9e1fe33",
  "input": {
    "key": "group_confidence_level",
    "value": [null]
  }
}

```

### Checklist


### Further comments

Now that the new schema definition is in place, the job done on the subject of user's confidence level has been fully reworked.
Thus this PR invalidates the previous PR on the subject: #5283 #5292 #5323
